syntax/parse.go

```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
 	}
 	return false
 }
+
+// Comment assignment.
+// We build two lists of all subexpressions, preorder and postorder.
+// The preorder list is ordered by start location, with outer expressions first.
+// The postorder list is ordered by end location, with outer expressions last.
+// We use the preorder list to assign each whole-line comment to the syntax
+// immediately following it, and we use the postorder list to assign each
+// end-of-line comment to the syntax immediately preceding it.
+
+// flattenAST returns the list of AST nodes, both in prefix order and in postfix
+// order.
+func flattenAST(f *File) (pre, post []Node) {
+	stack := []Node{}
+	Walk(f, func(n Node) bool {
+		if n != nil {
+			pre = append(pre, n)
+			stack = append(stack, n)
+		} else {
+			post = append(post, stack[len(stack)-1])
+			stack = stack[:len(stack)-1]
+		}
+		return true
+	})
+	return pre, post
+}
+
+// assignComments attaches comments to nearby syntax.
+func (p *parser) assignComments(f *File) {
+	pre, post := flattenAST(f)
+
+	// Split into whole-line comments and suffix comments.
+	var line, suffix []Comment
+	for _, com := range p.in.comments {
+		if com.Suffix {
+			suffix = append(suffix, com)
+		} else {
+			line = append(line, com)
+		}
+	}
+
+	// Assign line comments to syntax immediately following.
+	for _, x := range pre {
+		start, _ := x.Span()
+		x.AllocComments()
```

Comment:
Don't call AllocComments until you need to!

Reply:
Done